[WIP] Add extra optimizer stage

randomnetcat opened the pull request:
Closes #4674.

Very WIP. Please do not merge.

randomnetcat committed:
Initial stages

randomnetcat committed:
more changes

chriseth commented on libevmasm/ThirdOptimizerTesting.cpp:
```diff
@@ -0,0 +1,105 @@
+#include <algorithm>
+
+
+bool dev::solidity::Pattern::matches(std::vector<AssemblyItem> _items) const
```
The other two implementations of `matches` receive a single expression, so this would correspond to a `std::iterator` or an offset inside the list .This means that the first loop below would be outside of this function in the optimizer stage and not in the pattern. But if this is easier for you to work with for now, it's of course fine.

chriseth reviewed (commented):
Building up a stack and creating a pattern sounds rather complicated. It might be easier to write a function in the following way (but I might be totally wrong and you might actually really need a stack):
```
// Returns whether it matches or not and the size of the match
std::pair<bool, size_t> matches(vector<AssemblyItems> const& _items, size_t _offset)
{
  switch (kind()) {
  case Kind::Constant:
    // TODO: also handle the "any constant" case
    return make_pair(_items[_offset] == constant(), 1);
  case Kind::Operation:
    // check if the operation itself matches and
    // if the last argument matches at offset - 1.
    // if yes, get its size and check that the
    // argument before that matches before it etc.
  }   
}
```

Note that applying the pattern means doing replacements in the assembly item vector itself.

Also note that this will not work for many cases, but for some.

randomnetcat commented:
The reason I was building it into a tree was that the opcode stream is inherently tree-like. It's easy enough to do it with a flat list if there is a definite series of opcodes / constants to match, like {1, 1, MUL}, but as soon as there is just a kind of pattern, like match any operation, or match any operation or push constant, then there is no longer a definite length to look for. This means the code wouldn't know how far back to generate an expression for.

The main issue with the current method (that I can think of so far) is that it ignores things like JUMPDEST that don't modify the stack at all. My idea was then going to be to only look at sections with instructions that don't have any side effects, so any proven-correct optimizations could happen regardless of how they modified the stream.

**EDIT:** scratch that just thought of how to do it without a stack

randomnetcat commented:
updated to not parse items into a stack

randomnetcat committed:
changes
actually partially maybe kind of works now?

randomnetcat committed:
Only run if expression meets some criteria
Criteria: no instructions with >1 return value (too complicated for this module) & no side effects

chriseth commented on libevmasm/ThirdOptimizerTesting.h:
```diff
@@ -0,0 +1,180 @@
+
+		Kind m_kind;
+
+		bool m_hasConstant = false;
```
This screams for `boost::optional<u256> m_constant` ;)

chriseth commented on libevmasm/ThirdOptimizerTesting.cpp:
```diff
@@ -0,0 +1,279 @@
+	return currentEnd;
+}
+
+vector<vector<AssemblyItem>> parseExpressions(vector<AssemblyItem> const& _items, signed _limit = -1)
```
This looks like rather complicated code and it seems to do a lot of copying.
Do you think it might be possible for the `pattern::match` function to just forward the appropriate offset adjustments to its arguments?

chriseth commented on libevmasm/ThirdOptimizerTesting.cpp:
```diff
@@ -0,0 +1,279 @@
+	{
+		case NewOptimizerPattern::Kind::Any:
+			assertThrow(_pattern.isBound(), OptimizerException, "cannot create items from non-bound pattern");
+			return _pattern.boundItems();
```
Shouldn't this function first check whether it has bound items and return them in any case (and only then go into the switch)?

chriseth commented on libevmasm/ThirdOptimizerTesting.cpp:
```diff
@@ -0,0 +1,279 @@
+				vector<AssemblyItem> instructions{};
+
+				for (auto it = operands.rbegin(); it != operands.rend(); it++)
+					for (auto const& item : createItems(*it))
```
`instructions += createItems(*it)`

chriseth reviewed (changes requested):
This is a great start, but I think it should be simplified a little and it could also be improved with regards to efficiency.

I think the bound items should not be a copy and it might also make sense to just pass the to be filled vector to `createItems`, where it then appends.

chriseth commented:
Ah and concerning jumpdest: If you encounter anything that is not a simple operation (perhaps with `dup` being a future exception), you should not continue.

randomnetcat commented:
So, instead of just not applying any rules to that expression as a whole, bail out entirely?

randomnetcat committed:
Use boost::optional instead of field/has pair

chriseth commented:
@meowingtwurtle Things like `... jumpi push 4 push 5 add ` should still be simplified to `... jumpi push 9`.

chriseth commented:
But you should not do any analysis on `push 4 jumpdest push 5 add`

randomnetcat committed:
Always return bound items from createItems if bound

randomnetcat commented:
That's what the code currently tries to do, although maybe not quite completely. If the expression as a whole contains operations that can't be handled, it tries to optimize each of the arguments to that expression.

randomnetcat committed:
Replace a push_back loop with operator from CommonData

randomnetcat committed:
Fix bug in parseArguments
Now also more efficient

randomnetcat committed:
Add support for instructions without return values

randomnetcat commented:
In the first example you gave, if the add expression will be optimized and replaced. In the second example, the arguments to add will be parsed as `push 5` and `push 4 jumpdest`. If `push 5` could be optimized, it would. `push 4 jumpdest` as a whole wouldn't be optimized. But, now that I think of it, if `push 4` could be optimized, it should be.

chriseth commented:
Please try to keep this as simple as possible in the first iteration. If it only applies to very very specific cases, it is still an improvement. The most important part is that it is easy to see that it is perfectly correct. If that is done and working, we can try to extend it to more cases.

chriseth commented on libevmasm/ThirdOptimizerTesting.h:
```diff
@@ -0,0 +1,177 @@
+		boost::optional<AssemblyItemType> m_assemblyType;
+	};
+
+	std::shared_ptr<Underlying> m_ptr;
```
What is the purpose of this?

randomnetcat replied:
To create reference semantics. Patterns for operation operands have to be copied in to the parent (otherwise we could reference a pattern outside its lifetime). If we were to use plain value-fields, since they are copied in, in RulesList, the pattern that matches the opcodes will have copies of X and Y, which means the copies will be bound, and the replacement lambda captures by value, so the pattern matching (which is the one that is bound) and the replacement (which is what items are created from) are disjoint.

This means that once the pattern is bound in matching, the replacement lambda's version will also be bound.

chriseth replied:
The other implementations solve this via `matchGroups`, but your version might be more elegant. Have you considered using `shared_ptr`s instead? I.e. put all the members directly into the class but reference operands through shared pointers.

randomnetcat replied:
That would be equivalent, just with more reference counting, so it's up to style. If you want that change, it's fine.

randomnetcat replied:
Oh, might have misread that. Are you suggesting only operands by under a shared_ptr?

chriseth replied:
I'm suggesting moving the members into the outer class and using explicit shared_ptrs everywhere. The benefit would be that you would not have to use `m_ptr-><member>` all the time and the constructor would also be way simpler.

randomnetcat replied:
So make every field a direct shared_ptr member instead of having a shared_ptr to a struct?

chriseth replied:
No, just make every use of `NewOptimizerPattern` a `shared_ptr<NewOptimizerPattern>`

randomnetcat replied:
Okay.

randomnetcat replied:
Well, the way it is used in RuleList would require a wrapper class around shared_ptr that provides a d() method

chriseth commented on libevmasm/ThirdOptimizerTesting.cpp:
```diff
@@ -0,0 +1,313 @@
+	if (_unbind) unbind();
+
+	if (isBound())
+		return boundItems() == _items;
```
I am not sure if this is enough. The items have to be identical under any circumstance, and not just consist of the same opcodes.

randomnetcat replied:
boundItems consists of all of the items, not just opcodes. If AssemblyItem's operator== compares data (which it does), then this should check if items are exactly identical.

chriseth replied:
Sure, but what about e.g. `dup3` or `sload`? Both depend on the context and just comparing them is not enough.

randomnetcat replied:
`dup3` would only exist with arguments, so that would be fine.  `sload` is more tricky, but since it has side effects, it would never have rules applied against it in this module. And, it would only match if both the location and value were the same.

chriseth commented on libevmasm/ThirdOptimizerTesting.cpp:
```diff
@@ -0,0 +1,313 @@
+			assertThrow(instruction() == _items.back().instruction(), OptimizerException, "invalid bind instruction");
+	}
+
+	m_ptr->m_boundItems = _items;
```
This should not copy, if possible.

randomnetcat replied:
Well, it's passed by const ref, so a copy is necessary here. The other option would be to pass by value and move into place.

chriseth replied:
`m_boundItems` is not a reference, so it will copy.

randomnetcat replied:
I meant `_items`. I'm not sure `m_boundItems` can be a reference, that seems like it might cause lifetime issues, but I could be totally wrong.

chriseth replied:
Indeed, it cannot be a reference, that's why I would propose to work with offsets into the existing item vector.

randomnetcat replied:
If we were to use offsets there, we would have to use offsets/iterators basically everywhere, since right now there is not one singular items vector, but a bunch of vectors copied from each other.

randomnetcat committed:
Use iterators everywhere feasible
Yay for less copying!

randomnetcat commented:
@chriseth Just reduced copying by a lot with iterators. Valgrind reports no memory leaks or invalid reads.

chriseth commented on libevmasm/ThirdOptimizerTesting.cpp:
```diff
@@ -0,0 +1,368 @@
+}
+
+template<typename It, typename>
+bool dev::solidity::NewOptimizerPattern::matches(It _begin, It _end, bool _unbind)
```
Why does it need to match a range? Why can't it just start trying to match at a certain point and then work its way backwards?

chriseth commented on libevmasm/ThirdOptimizerTesting.cpp:
```diff
@@ -0,0 +1,368 @@
+
+			auto const instruction = this->instruction();
+
+			if (back.type() != AssemblyItemType::Operation) return false;
```
This can go to the very beginning of the case.

chriseth commented on libevmasm/ThirdOptimizerTesting.cpp:
```diff
@@ -0,0 +1,368 @@
+			if (back.type() != AssemblyItemType::Operation) return false;
+			if (back.instruction() != instruction) return false;
+
+			auto const argumentCount = instructionInfo(instruction).args;
```
Please reduce the amount of `auto` in your code, it makes it very hard to read.

chriseth commented on libevmasm/ThirdOptimizerTesting.cpp:
```diff
@@ -0,0 +1,368 @@
+	return expressions;
+}
+
+template<typename It, typename = typename std::enable_if<is_assembly_item<decltype(*std::declval<It>())>::value>::type>
```
Is the purpose of this to work both for `const` and `non-const`?

randomnetcat replied:
Well, the purpose of this line specifically is to ensure that no programmer tries to pass in some random iterator. `is_assembly_item` is defined such that it can be any const/non-const iterator.

chriseth commented on libevmasm/ThirdOptimizerTesting.cpp:
```diff
@@ -0,0 +1,368 @@
+	auto expressions = parseExpressions(_begin, _end - 1, neededArguments, false);
+	assertThrow(expressions.size() >= neededArguments, OptimizerException, "more arguments needed than provided");
+
+	expressions = vector<It>(expressions.begin(), expressions.begin() + neededArguments);
```
Why do you need to return a vector instead of a pair of iterators?

randomnetcat replied:
That would be equally valid, but iterators are (hopefully) cheap to copy. And, an iterator to iterator is just a little bit confusing.

randomnetcat replied:
Oh, and the vector is dead at that point, so you would have iterators into a destructed vector.

chriseth reviewed (commented):
Perhaps consider using boost::range ( https://www.boost.org/doc/libs/1_68_0/libs/range/test/sub_range.cpp ) or declaring a struct with two iterators instead of two parameters for begin/end.

chriseth reviewed (commented):
Ah sorry, I see that you are already using `iterator_pair` at some places. I think it should be used everywhere where two iterators represent a range (if it makes sense) to claify that they belong together.

I still don't see why `backExpressionBegin` and all the other helper functions are necessary here. Perhaps they could all be removed if `match` would be implemented in the following way:

It is basically built in the same way except that it does not return a `bool` but instead an iterator. If it returns `_end`, then it is basically a no-match. If it returns something else, the pattern matched everything in the range from its return value until `_end`. It would be called with `_begin` being the very beginning of the list of assembly items, and `_begin` would only serve as a marker for where to stop. The match has to end at `_end`, so it does not really search.

Now inside the implementation of `match`, you check if the operation at `*_end` is of the correct type. You get the number of arguments of the instruction and then you recursively call match:

```
iterator it = _end - 1;
for (int i = numArguments - 1; i >= 0; --i) {
  iterator m = m_operands[i].match(it);
  if (m == it) return _end;
  it = m;
}
return it;
```

randomnetcat committed:
Use iterator_pair everywhere

randomnetcat committed:
Move addDefaultRules call to constructor

randomnetcat commented:
There are two reasons that those helpers exist. The first is that they can be implemented in terms of each other and recursively. The second is that they can be used to decompose a larger opcode stream that is fed in. This means that `optimize` can break down an entire opcode stream into its expressions and arguments. Having those functions allows the code to be reused.

randomnetcat committed:
Make requested trivial changes
Document helper methods

randomnetcat commented:
@chriseth Made the trivial requested changes (everything requested except explicit shared_ptr usage). I also documented the helper methods, hopefully that should help you and others understand them.

axic commented:
@meowingtwurtle @chriseth when/where did we lose track of this one?

randomnetcat commented:
A few things: chriseth believes that the current code in this is too complicated (I somewhat agree, but I'm not sure how much simpler it could be) and this was informally returned to the public (see the discussion on the main issue page).

chriseth commented:
I'm closing this. I would propose to award half of the bounty to meowingtwurtle for the work he put into this (-> discussion on the issue).
